pennylane/transforms/diagonalize_measurements.py

```diff
@@ -0,0 +1,338 @@
+"""Transform to diagonalize measurements on a tape, assuming all measurements are commuting."""
+
+from copy import copy
+from functools import singledispatch
+
+import pennylane as qml
+from pennylane.operation import Tensor
+from pennylane.ops import CompositeOp, LinearCombination, SymbolicOp
+from pennylane.transforms.core import transform
+
+# pylint: disable=protected-access
+
+_default_supported_obs = (qml.Z, qml.Identity)
+
+
+def null_postprocessing(results):
+    """A postprocessing function returned by a transform that only converts the batch of results
+    into a result for a single ``QuantumTape``.
+    """
+    return results[0]
+
+
+@transform
+def diagonalize_measurements(tape, supported_base_obs=_default_supported_obs):
+    """Diagonalize the measurements on a tape if they are not supported. Raises an error if the
+    measurements do not commute.
```

Comment:
```suggestion
    """Diagonalizes any unsupported observables from the local Pauli basis. Raises an error if the measurements do not commute.
```

Reply:
@trbromley , does the wording on this sound alright? And is this enough, or would a `..note` be better?

Reply:
```suggestion
    """Diagonalize a set of measurements into the standard basis.
    
    See the usage details for more information on which
    measurements are supported.
```
I'm thinking it's easier to leave the details to the usage details section.